presto-main/src/main/java/com/facebook/presto/operator/aggregation/differentialentropy/StateSerializer.java

```diff
@@ -0,0 +1,282 @@
+                throw new PrestoException(
+                        INVALID_FUNCTION_ARGUMENT,
+                        "Inconsistent strategy");
+            }
+            ((WeightedReservoirSampleStateStrategy) target).mergeWith((WeightedReservoirSampleStateStrategy) source);
+            return;
+        }
+
+        if (target instanceof FixedHistogramMLEStateStrategy) {
+            if (!(source instanceof FixedHistogramMLEStateStrategy)) {
+                throw new PrestoException(
+                        INVALID_FUNCTION_ARGUMENT,
+                        "Inconsistent strategy");
+            }
+            ((FixedHistogramMLEStateStrategy) target).mergeWith((FixedHistogramMLEStateStrategy) source);
+            return;
+        }
+
+        if (target instanceof FixedHistogramJacknifeStateStrategy) {
+            if (!(source instanceof FixedHistogramJacknifeStateStrategy)) {
+                throw new PrestoException(
+                        INVALID_FUNCTION_ARGUMENT,
+                        "Inconsistent strategy");
+            }
+            ((FixedHistogramJacknifeStateStrategy) target).mergeWith((FixedHistogramJacknifeStateStrategy) source);
+            return;
+        }
+
+        throw new InvalidParameterException("unknown strategy combination");
+    }
+
+    @Override
+    public Type getSerializedType()
+    {
+        return VARBINARY;
+    }
+
+    @Override
+    public void serialize(State state, BlockBuilder out)
+    {
+        final int requiredBytes =
+                SizeOf.SIZE_OF_INT + // Method
+                        (state.getStrategy() == null ? 0 : state.getStrategy().getRequiredBytesForSerialization());
+
+        final StateStrategy strategy = state.getStrategy();
+
+        SliceOutput sliceOut = Slices.allocate(requiredBytes).getOutput();
+
+        if (strategy == null) {
+            sliceOut.appendInt(0);
+        }
+        else if (strategy instanceof UnweightedReservoirSampleStateStrategy) {
+            sliceOut.appendInt(1);
+        }
+        else if (strategy instanceof WeightedReservoirSampleStateStrategy) {
+            sliceOut.appendInt(2);
+        }
+        else if (strategy instanceof FixedHistogramMLEStateStrategy) {
+            sliceOut.appendInt(3);
+        }
+        else if (strategy instanceof FixedHistogramJacknifeStateStrategy) {
+            sliceOut.appendInt(4);
+        }
+        else {
+            throw new InvalidParameterException("unknown method in serialize");
+        }
+
+        if (strategy != null) {
+            strategy.serialize(sliceOut);
+        }
+
+        VARBINARY.writeSlice(out, sliceOut.getUnderlyingSlice());
+    }
+
+    @Override
+    public void deserialize(
+            Block block,
+            int index,
+            State state)
+    {
+        StateSerializer.deserialize(
+                VARBINARY.getSlice(block, index).getInput(),
+                state);
+    }
+
+    public static void deserialize(
```

Comment:
- inline this method
- put arguments on the same line
- use verify or make error message user-friendly

```
    @Override
    public void deserialize(Block block, int index, State state)
    {
        StateStrategy strategy = deserializeStrategy(VARBINARY.getSlice(block, index).getInput());
        if (strategy == null && state.getStrategy() != null) {
            // TODO Verify or Throw?
        }
        if (strategy != null) {
            state.setStrategy(strategy);
        }
    }
```